examples/soroban/nft/src/nft.rs

```diff
@@ -0,0 +1,97 @@
+use loam_sdk::{
+    soroban_sdk::{self, contracttype, env, Address, Bytes, Lazy, Map},
+    IntoKey,
+};
+use loam_subcontract_core::Core;
+
+use crate::{
+    subcontract::{IsInitable, IsNonFungible},
+    Contract,
+};
+
+#[contracttype]
+#[derive(IntoKey)]
+pub struct MyNonFungibleToken {
+    admin: Address,
+    name: Bytes,
+    total_count: u32,
+    owners_to_nft_ids: Map<Address, u32>,
+    nft_ids_to_owners: Map<u32, Address>,
+    nft_ids_to_metadata: Map<u32, Bytes>,
+}
+
+impl MyNonFungibleToken {
+    #[must_use]
+    pub fn new(admin: Address, name: Bytes) -> Self {
+        MyNonFungibleToken {
+            admin,
+            name,
+            total_count: 0,
+            owners_to_nft_ids: Map::new(env()),
+            nft_ids_to_owners: Map::new(env()),
+            nft_ids_to_metadata: Map::new(env()),
+        }
+    }
+}
+
+impl Default for MyNonFungibleToken {
+    fn default() -> Self {
+        MyNonFungibleToken::new(env().current_contract_address(), Bytes::new(env()))
+    }
+}
+
+impl IsInitable for MyNonFungibleToken {
+    fn nft_init(&mut self, admin: Address, name: Bytes) {
```

Comment:
```suggestion
    fn nft_init(&self, admin: Address, name: Bytes) {
```
If it's `&mut` it does the following in the caller.
```rust
let mut s = Initable::Self::get_lazy().unwrap_or_default();
let res = s.nft_init(&mut s, admin, name);
Initable::Self::set_lazy(s);
res
```
So in this case it might be overwritten with the default. This does suggest that we have a bug in the implementation of init methods. Ideally, the value should initially be `None` so that calling it twice wouldn't overwrite the value. I would need to change the design and macro. But the more I think about it. It should be that in the init case it should be `None` and for all other methods it should be `Some` as otherwise means the contract hasn't been initialized.

Reply:
This shouldn't be an issue if we ensure that when say using smartdeploy it's initialized in the same transaction as its deployed.

Reply:
I think I understand  - `&mut self` here opens us up to the possibility that the internal state of the singleton can be overwritten. 

Are you saying that you want to redesign this so that users aren't able to get themselves into that situation by initializing a contract with a state of `None`. And then other methods require that the state is `Some`? Otherwise it will indicate that the contract hasn't been initialized. 

That all makes sense to me!

Reply:
Also, i think that we have a `&mut self` in the fungible token example as well - does it make sense to update that too?

Reply:
updated the fungible token example here: 80eba5b6adb9f7984eaeda4ddb51691909d0c088

Reply:
Yep! We should add and `init` decorator! I'll open an issue.